pkg/cmd/pulumi/templates/templates.go

```diff
@@ -0,0 +1,227 @@
+	s.errors = append(s.errors, err)
+	s.m.Unlock()
+}
+
+func (s *Source) addErrorOnEmpty(err error) {
+	s.lockOpen("add error on empty")
+	s.errorOnEmpty = append(s.errorOnEmpty, err)
+	s.m.Unlock()
+}
+
+func (s *Source) lockOpen(action string) {
+	s.m.Lock()
+	if s.closed {
+		panic("Attempted to act on closed source: " + action)
+	}
+}
+
+// Close cleans up the [Source] and any associated templates.
+//
+// Close should always be called when [Source] is dropped.
+func (s *Source) Close() error {
+	s.wg.Wait() // Wait to ensure that all templates have been fetched so all closers are visible.
+
+	s.lockOpen("close")
+	defer s.m.Unlock()
+	s.closed = true
+	errs := make([]error, len(s.closers))
+	for i, f := range s.closers {
+		errs[i] = f()
+	}
+	return errors.Join(errs...)
+}
+
+type Template struct {
+	name               string
+	description        string
+	projectDescription string
+
+	error error
+
+	download func(ctx context.Context) (workspace.Template, error)
+
+	source *Source
+}
+
+func (t Template) Name() string               { return t.name }
+func (t Template) Description() string        { return t.description }
+func (t Template) ProjectDescription() string { return t.projectDescription }
+func (t Template) Error() error               { return t.error }
+
+func (t Template) Download(ctx context.Context) (workspace.Template, error) {
+	if t.source == nil {
+		panic("Cannot download a template without a host")
+	}
+	if t.source.closed {
+		panic("Cannot download a template from an already closed host")
+	}
+
+	return t.download(ctx)
+}
+
+type Scope struct{ kind string }
+
+var (
+	ScopeDefault     = Scope{}
+	scopeDefault     = ScopeAll
+	ScopeTraditional = Scope{"traditional"}
+	ScopeLocal       = Scope{"none"}
+	ScopeAll         = Scope{"all"}
+)
+
+// Create a new [Template] [Source] associated with a given [Scope].
+func New(
+	ctx context.Context, templateNamePathOrURL string, scope Scope,
+	templateKind workspace.TemplateKind, interactive bool,
+) *Source {
+	// apply the default scope, if necessary
+	if scope == ScopeDefault {
+		scope = scopeDefault
+	}
+
+	var source Source
+	ctx, cancel := context.WithCancel(ctx)
+	source.closers = append(source.closers, func() error { cancel(); return nil })
+
+	queryKind := getTemplateQuery(templateNamePathOrURL)
+
+	if scope == ScopeAll || scope == ScopeTraditional || scope == ScopeLocal {
+		source.wg.Add(1)
+		go func() {
+			source.getWorkspaceTemplates(ctx, templateNamePathOrURL, scope, templateKind, &source.wg)
+			source.wg.Done()
+		}()
+	}
+
+	if scope == ScopeAll && templateKind == workspace.TemplateKindPulumiProject && queryKind == queryName {
+		source.wg.Add(1)
+		go func() {
+			source.getOrgTemplates(ctx, templateNamePathOrURL, interactive, &source.wg)
+			source.wg.Done()
+		}()
+	}
+
+	return &source
+}
+
+type queryKind string
+
+const (
+	queryName queryKind = "query-name"
+	queryURL  queryKind = "query-url"
+	queryPath queryKind = "query-path"
+)
```

Comment:
some docstrings with an explanation why this is here and what it is doing would be helpful.

Reply:
Made irrelevent by https://github.com/pulumi/pulumi/pull/18627#discussion_r2007136988.